### DISCUSSION TITLE
Periodicity with relative phase

### DISCUSSION THREAD

**Discussion by HelgeGehring** [kinnala/scikit-fem#989] [Q&A]
If I have a mesh like
![image](https://user-images.githubusercontent.com/42973196/204634918-064aaf85-6c8e-4a27-8eb1-186ae6789a35.png)
and I want it to be periodic, i.e. left-right and top-bottom with the possibility to set a relative phase between the facets, how would I do that?

More precise: I want the DOFs at the left side to be the same as at the right side, but multiplied by something like `i`, which I'd like to define.

I'd guess I need for each side a facet basis and add a term, right? But I don't have a clue how that should look like :/ Would be awesome if you could give me some advise how to get started :)

**Comment by kinnala** [comment]
I'd start by trying to add a penalty term and then try to improve from there. I can try to make you an example.

**Reply by HelgeGehring** [reply]
Awesome, thanks! (I'll use it in the end with a gmsh generated mesh, but the edges will also be parallel/same length)

**Reply by kinnala** [reply]
So you want to set a constraint like
$$u(0,y) = 2 u(1,y)$$
This can be done using a penalty term such as
$$\frac{1}{\epsilon}(u(0,y) - 2u(1,y), v(0,y) - 2v(1,y))$$

**Reply by kinnala** [reply]
Here is the code for such penalty term:
```python
from skfem import *
from skfem.models import laplace, unit_load

m = MeshTri().refined(4)
e = ElementTriP1()

basis = Basis(m, e)

fbases = [
    FacetBasis(m, e, facets='left'),
    FacetBasis(m, e, facets='right'),
]
# this relies on the facets 'left' and 'right' being returned by Mesh.facets_satisfying in the "same" order
# if this doesn't happen then additional sorting step needs to be done (TODO)

@BilinearForm
def penalty(u, v, w):
    u1 = (w.idx[0] == 0) * u
    u2 = (w.idx[0] == 1) * u
    v1 = (w.idx[1] == 0) * v
    v2 = (w.idx[1] == 1) * v
    ju = u1 - 2 * u2
    jv = v1 - 2 * v2
    return 1. / 1e-2 * ju * jv

A = asm(laplace, basis)
B = asm(penalty, fbases, fbases)
f = asm(unit_load, basis)

y = solve(*condense(A + B, f, D=basis.get_dofs({'top', 'bottom'})))

basis.plot(y, colorbar=True, shading='gouraud', levels=5).show()
```
![Figure_1](https://user-images.githubusercontent.com/973268/204897207-a4ed079e-c841-420d-9eb2-f9ef23a854d4.png)

I'm sorry for the hacky form definition, I never foresaw anyone wanting to do this. Perhaps we need to add something in ´skfem.helpers`.

**Reply by kinnala** [reply]
Note my comment about the sorting of the facets. Those two `FacetBasis` objects are not compatible unless the facets are in the correct order and this is not verified by the above code.

**Reply by kinnala** [reply]
This oneliner checks if the facets are in the same order:
```python
fbases[0].default_parameters()['x'][1] == fbases[1].default_parameters()['x'][1]
```
Fixing it in case they aren't requires some sorting.

**Reply by HelgeGehring** [reply]
Thank you so much! I'll try it out!

**Reply by vr0110** [reply]
> I never foresaw anyone wanting to do this

I'm also very interested in simulating these kinds of periodic structures and it's very helpful for me to study this example. So thank you both!

It's actually a common problem in the analysis of propagation modes in periodic structures such as crystals and meta materials.

See here for more information (plots above and below this bookmark):
https://mpb.readthedocs.io/en/latest/Python_Data_Analysis_Tutorial/#the-source-of-the-tm-gap-examining-the-modes

**Reply by kinnala** [reply]
```python
from skfem import *
from skfem.helpers import *
from skfem.models import laplace, unit_load

m = MeshTri().refined(4)
e = ElementTriP1()

basis = Basis(m, e)

fbases = [
    FacetBasis(m, e, facets='left'),
    FacetBasis(m, e, facets='right'),
]

alpha = 1e-2

def split(*args):
    m = len(args) - 2
    ns = args[-1]
    w = args[-2]
    return tuple(
        tuple(args[j] if w.idx[j] == i else args[j].zeros() for i in range(ns[j]))
        for j in range(m)
    )

@BilinearForm
def penalty(u, v, w):
    (u1, u2), (v1, v2) = split(u, v, w, [2, 2])
    ju = u1 - 2 * u2
    jv = v1 - 2 * v2
    mu = .5 * (dot(grad(u1), w.n) + 1 / 2 * dot(grad(u2), w.n))
    mv = .5 * (dot(grad(v1), w.n) + 1 / 2 * dot(grad(v2), w.n))
    return 1. / (alpha * w.h) * ju * jv - ju * mv - mu * jv

params = {
    'n': fbases[0].normals,  # normals from one side only
}

A = asm(laplace, basis)
B = asm(penalty, fbases, fbases, **params)
f = asm(unit_load, basis)

y = solve(*condense(A + B, f, D=basis.get_dofs({'top', 'bottom'})))

basis.plot(y, colorbar=True, shading='gouraud', levels=5).show()
```
Here I tried to work on this `split` helper. I also tried implementing Nitsche's method instead of penalty method.

**Reply by HelgeGehring** [reply]
Thanks! I've started using it in https://github.com/HelgeGehring/femwell/blob/main/femwell/mode_solver_inplane.py , but it doesn't yet give the results I expect - still working on it. Would you have any concerns using it with complex numbers?

And could you maybe add a few comments what the code actually does? I have a hard time understanding it. I'd guess a few minutes from you could save me some hours for finding the right reference :D (my background is physics and I'm currently fighting my way into FEM. By doing all this photonics/electronics/thermal simulations and your help I learn a lot, but this is way more complicated) To me it seems to add a term which is proportional to the difference at the interfaces, right? Does it need to have the vertices on the two facets to be spaced the same?
Nitsche's method would probably be exact and prevent issues?

Besides that, I'd like to share the tools I develop with some more coworkers. For this a new skfem version would be nice, as it simplifies the Installation process for them a lot. Would you mind releasing somewhen soonish a new version? :)

@vr0110 that's exactly what I'm working on! Check out the file I've linked, would be happy if you could join the development of these photonic tools :)

**Reply by kinnala** [reply]
Version 8.0.0 is now released. The above code requires the vertices to match. However, penalty or Nitsche's method in general does not require the vertices to match.  If the vertices are not matching there needs to be yet another integration mesh with the vertices from the both sides and index mapping from the integration mesh to the original meshes.

There are also other approaches for this type of constraint. If you know that the vertices are matching and can find the index mapping between matching vertices then you can augment the linear system with Lagrange multipliers that force the DOF values to be equal (+ constant). There might be also some other ways to impose the constraint, I think looking for "multi-point constraints" might lead to relevant results.

**Reply by kinnala** [reply]
Here is how to do it using the index mapping and discrete Lagrange multipliers:
```python
import numpy as np
from skfem import *
from skfem.models import laplace, unit_load

m = MeshTri().refined(4)
e = ElementTriP1()

basis = Basis(m, e)

left = basis.get_dofs(facets='left').nodal['u']
right = basis.get_dofs(facets='right').nodal['u']
D = basis.get_dofs({'top', 'bottom'}).all()

left = np.setdiff1d(left, D)
right = np.setdiff1d(right, D)

B = np.zeros((len(left), basis.N))
B[range(len(left)), left] = 1
B[range(len(right)), right] = -2

A = asm(laplace, basis)
f = asm(unit_load, basis)

A = bmat([[A, B.T],
          [B, None]], 'csr')
f = np.concatenate((f, 0 * left))

y = solve(*condense(A, f, D=basis.get_dofs({'top', 'bottom'})))

basis.plot(y[:basis.N], colorbar=True, shading='gouraud', levels=5).show()

```

**Reply by HelgeGehring** [reply]
Thanks for the new version!

And thanks for the examples! It seems to start working for me :)
If I understand it right, the penalty method adds a term which forces the DOFs to be equal (where a wrong choice of `alpha` would lead to problems (?)) and the Lagrange multipliers increase the matrix size by the number of DOFs which have to match and use the multipliers to force them to match.

"multi-point constraints" helped a lot for searching, thanks! 
The Master Slave Method (see https://mashayekhi.iut.ac.ir/sites/mashayekhi.iut.ac.ir/files//files_course/lesson_16.pdf) would seem to be the ideal solution, right? It would be exact and would also decrease the size of the system. The implementation would probably be somehow similar to condense, where some DOFs are kicked out of the system and subsequently are derived from the solutions (?)

**Reply by kinnala** [reply]
If you come up with something like `condense` which deals with these multi-point constraints, I'd be happy to accept it as a PR to `skfem.utils`.

**Reply by kinnala** [reply]
If I understand your reference correctly:
- `T` and `g` specify the constraint and are thus given by the user.
- So we could have a call such as `multipoint(K, f, T=T, g=g)` which returns the modified system and can be directly passed to `solve`.

Now the user probably also wants to set boundary conditions by specifying `D` and `I` and `x` etc. I think these can be technically implemented by specific choices of `T` and `g` so perhaps we want to have "helper" functions for creating `T` and `g` for (a) boundary conditions (b) periodic conditions (c) possible other interesting scenarios.

**Reply by HelgeGehring** [reply]
That's exactly how I understand it! Seems like a more generalized form of `condense`.

I'm unsure what exactly the function signature of the helper functions should look like. In principal also a three or more point constraints would be doable like this. Also I'm thinking of mixing different conditions, i.e. two sides periodic and setting the other two sides to 0.

If we implement the "Nonhomogeneous MPCs", that should already cover (a) and (b).

Maybe something like
```
def constraint(Dx, x, *Ix): -> T,g
```
where `Dx=(D,x)` and `D` are the degrees of freedom to eliminate, `x` is the prefactor. `Ix=(I,x)` would be similar defined (and multiple `Ix` and also None can be put in) and `x` would be like in condense the `x`?
I know that the `x` in `Dx` would not be completely necessary, as it can be realized by the choice of Ix and g, but I'd guess that makes it way more convenient and better understandable.

Probably we would also want a function which can combine several constraints into one `T,g` (and checks their compatibility)?
This could either be done before creating `T,g` by combining Dx, x and Ix or after if there's a way to combine already created `T,g` efficiently?

**Reply by HelgeGehring** [reply]
see #1006

**Comment by iceneve** [comment]
from :  谢志博 ***@***.***
